### ISSUE TITLE
[FEATURE] [RFC] Integrations Dynamic Catalogs

### ISSUE THREAD

**Issue by Swiddis** [opensearch-project/dashboards-observability#1253]
## Is your feature request related to a problem?
In 2.9, the Integrations project was released with the ability to provide insights on data residing in a customer’s OpenSearch cluster. In 2.11, the functionality was extended to further allow reading raw log data on S3. However, the current integrations catalog is hard-coded in the application, meaning that changes to integration content are coupled with the OS release cadence. During both releases, there was a large rush to get content prepared for new features near or after code freeze, adding strain to the operations process and leading to bugs regarding the content released under time pressure.

## What solution would you like?
To remedy this, we propose decoupling the release of integrations content from the OpenSearch release process. We do this by introducing a Dynamic Catalog with the ability to install content online, which isn’t already shipped with the OpenSearch distribution. This also allows users to deploy their own proprietary catalogs. Some basic requirements and prototype designs are shared below, but a finalized design document is still in progress.

### Requirements

1. As a user, I want to update an integration with a developer fix if I’m experiencing problems with it, without needing to wait for a new release.
2. As a user, when I update a large set of integrations, I want to have access to historic versions of integrations which I may not want to use updated versions for.
3. As a user, I want to use integrations developed by third parties that may not have uploaded their work to OpenSearch repositories.
4. As a user, I want to receive updates for integrations without needing to manually reinstall bundles.
5. As a user, when I add a catalog, it should be restricted to my tenancy.

### UI

While a lot of the details of which upload methods will be supported is up in the air, the general flow will be adding an install button to the Integrations page where a source can be provided.

https://github.com/opensearch-project/dashboards-observability/assets/31739405/260267b0-bdbd-4b87-b3bb-b2714ab148bb

### Design

Much of the design is still under construction, I mainly intend to share some of the ideas for how I plan on implementing it.

#### Catalog Adaptors

As part of 2.10, we refactored integration loading to introduce the idea of a Catalog Adaptor. An Adaptor is responsible for reading and validating integrations from some storage location. Currently, the only adaptor is a File System Adaptor, which reads integrations directly from the file system, and is used to read the hardcoded catalog shipped with OS today. There are more possible adaptors under consideration.

* Index Adaptor: Read integrations from an OpenSearch Index.
    * Pros: Well-supported with existing OS Client tools.
    * Cons: Expressing integrations within an OS document may be complicated, and considerably different compared to our current FS serialization.
* Network Adaptor: Read integrations from a remote location.
    * Pros: Extremely flexible, works very well with third-party integration vendors, allows automatic updates.
    * Cons: Standardizing an API may be complicated, and this likely has heavy security considerations. Can’t stand on its own to stay compatible with air-gapped setups.
    * We can also provide multiple versions of this for different APIs, such as a GithubAdaptor that uses the GH Repo API.
* Zip File Adaptor: Read integrations from Zip files. Only used for serialization in indices.
    * Pros: Very portable, relatively simple to implement.
    * Cons: Might also have security implications.
* Json File Adaptor: Handled similarly to Zip files architecturally, and uses the same format as we use for reading from an index.

#### Serialization

Today, when a read request is routed to the `IntegrationsManager`, it currently uses a `TemplateManager` to scan for existing integrations. A `TemplateManager` in turn contains a list of `CatalogDataAdaptor`s that it uses to retrieve raw integration data, which it assembles and validates. Today, the `IntegrationsManager` is hardcoded with one `TemplateManager` set up for the current file system catalog.

To support dynamic catalogs, we will add an `IndexCatalogDataAdaptor`, along with In-Memory adaptors (such as Zip or Json). In-memory adaptors will only be used for re-serializing uploaded integrations into an index, and not . Re-serialized dynamic integrations are stored in an index, but there are some other sources, such as today’s `FileSystemAdaptor` for default shipped integrations, and a future `NetworkAdaptor` for auto-updating integrations. This flow is diagrammed below:

```mermaid
flowchart TD
    U(User)
    U -->|Requests| P[Integrations Router]
    subgraph pg[Integrations Plugin]
        P -->|Requests| M[[IntegrationsManager]]
        M -->|Compile Templates for Installation| B[[IntegrationBuilder]]
        M -->|Read/Install Integration Templates| I[[TemplateManager]]
        I -->|Install Templates| S("serialize()")
        I -->|Fetch Templates| cda

        subgraph cda[CatalogDataAdaptor]
            IMA[In-Memory Adaptors]
            FSA[[FileSystemAdaptor]]
            NWA[[NetworkAdaptor]]
            IXA[[IndexAdaptor]]
        end
    end
    M -->|Install Integrations| CC[[SavedObjectsClientContract]]
    CC --> IDX[(OS Indices)]
    S -->|Install Templates| CC
    FSA -->|Fetch Data| FS[(Built-in Files)]
    NWA -->|Fetch Data| NW((Network APIs))
    IXA -->|Fetch Data| CC
```

To re-serialize integrations, we need a flat format that keeps integrations in one document, unlike today's directory-based system. There are a few options for doing this, but a minimal-change one is to add an optional `content` field to integration config assets that accepts a base64-encoded file body, and modify adaptors to read from this field if it's available.

```json5
// POST /api/integrations/repository/upload
{
  "reader": "json",
  "content": {
    "object": {
      "name": "sample_integration",
      "version": "1.0.0",
      "displayName": "Sample Dashboard",
      "description": "Sample Dashboard",
      // ...
      "assets": {
        "savedObjects": {
          "name": "objects",
          "version": "1.0.0",
          "content": {
            "encoding": "base64",
            "data": "eyJoZWxsbyI6ICJ3b3JsZCJ9"
          },
        }
      }
      // ...
    }
  }
}
```

**What alternatives have you considered?**
N/A

**Do you have any additional context?**
Some historic child issues, future issues will link back to this:
- #1236 
- #1243

**Comment by Swiddis** [comment]
Questions and my current best guess on answers:

**Which exact adaptors should we implement?**
For 2.12 we’re at least targeting Zip, Json, and Index adaptors, while for 2.13 we're considering network-based adaptors like reading a GitHub repository.

**Where should we store catalogs long-term?**
I think they'll end up being an additional object within `.kibana`.

**How will we keep network adaptors stable in general? What tools/libraries can we use for network access? Will the user Need to adhere to a specific API or do we continually add adaptors (e.g. a GitHub adaptor using the GitHub dev API)?**
Will need to think more on this.

**Should we denote which catalog uploaded integrations are associated with? How do we find the catalog of uploaded integrations?**
There will need to be some way to distinguish duplicates (or detect/version them), but the details are still up in the air. A tag on the integration cards?

**Comment by kgcreative** [comment]
Couple of additional questions that come to mind:

1. I think I would really encourage us to use `upload` instead of `install` -- primarily because we do have tabs for `installed` & `available` already, and I would prefer that we don't confuse users with multiple meanings for `Install`
2. Can you talk a little bit more around the decision to name these `catalogs` as an abstraction, particularly if we do intend to launch a `catalog` that could include integrations, sample data sets, features, and other installable content. I feel like this could muddle the meaning of catalog prematurely, when what we are really doing is adding either installing single `integrations`, `integration sets`, or `bulk adding`

**Comment by Swiddis** [comment]
> I think I would really encourage us to use `upload` instead of `install`

I agree, I'll try to clarify the nomenclature as we get more of the flows documented and mocked.

> Can you talk a little bit more around the decision to name these catalogs as an abstraction, particularly if we do intend to launch a catalog that could include integrations, sample data sets, features, and other installable content.

Currently, the main feature differentiating a catalog from an integration bundle is that different catalogs can be serialized differently, and may in the future support different operations. For example, if we allow users to maintain remote catalogs in GitHub, those could feasibly support auto-updating, while file uploads would have a trickier time with that. There's a lot more detailed discussion of catalogs over on the [opensearch-catalog repo](https://github.com/opensearch-project/opensearch-catalog), which currently is a mirror of the content we add here, but with this feature will be itself a source for integrations.

I think that clarifying our vision for catalogs in the future will be important because you rightfully point out that it's kind-of a muddy concept right now. @YANG-DB has put a lot of work into clarifying its responsibilities and what it contains. I'd like to see his input.

**Comment by Swiddis** [comment]
Can be closed by #1381 and #1391 together

**Comment by Swiddis** [comment]
Or since there's a separate issue #1389 we can just close this with 1381
